samza-azure/src/main/java/org/apache/samza/AzureLeaderElector.java

```
@@ -0,0 +1,111 @@
+
+
+/**
+ * Class to facilitate leader election in Azure.
+ * The processor that acquires the lease on the blob becomes the leader.
+ * The lease ID is null initially. It is generated by Azure when the processor acquires the lease, and updated accordingly.
+ * Every processor requires a valid active lease ID in order to perform successful write and delete operations on the blob.
+ * Read operations from the blob are not dependent on the lease ID.
+ */
+public class AzureLeaderElector implements LeaderElector {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AzureLeaderElector.class);
+  private static final int LEASE_TIME_IN_SEC = 60;
+  private final LeaseBlobManager leaseBlobManager;
+  private LeaderElectorListener leaderElectorListener = null;
+  private AtomicReference<String> leaseId;
+  private AtomicBoolean isLeader;
+
+  public AzureLeaderElector(LeaseBlobManager leaseBlobManager) {
+    this.isLeader = new AtomicBoolean(false);
+    this.leaseBlobManager = leaseBlobManager;
+    this.isLeader = new AtomicBoolean(false);
+    this.leaseId = new AtomicReference<>(null);
+  }
+
+  @Override
+  public void setLeaderElectorListener(LeaderElectorListener listener) {
+    this.leaderElectorListener = listener;
+  }
+
+  /**
+   * Tries to become the leader by acquiring a lease on the blob.
+   * The acquireLease operation has a retry policy where upon failure, the operation is tried 3 times at 5 second intervals.
+   * Invokes the listener on becoming the leader.
+   */
+  @Override
+  public void tryBecomeLeader() {
+    try {
+      leaseId.getAndSet(leaseBlobManager.acquireLease(LEASE_TIME_IN_SEC, leaseId.get()));
+      if (leaseId.get() != null) {
+        LOG.info("Became leader!");
+        isLeader.set(true);
+        if (leaderElectorListener != null) {
+          leaderElectorListener.onBecomingLeader();
+        }
+      }
+    } catch (AzureException e) {
+      LOG.error("Error while trying to acquire lease.", e);
```

This exception occurs after a predefined number of retries according to the blob operations retry policy. Hence, I have swallowed the exception here and the Azure JC will take care of the corresponding behaviour.